File: CppSolution/SymmetricTree.cpp
#include <iostream>
#include <vector>
#include "Utility.h"
using namespace std;

class Solution
{
private:
    bool isSym(TreeNode *left, TreeNode *right)
    {
        if (left == NULL)
            return right == NULL;
        if (right == NULL)
            return left == NULL;
        if (left->val != right->val)
            return false;
        if (!isSym(left->left, right->right))
            return false;
        if (!isSym(left->right, right->left))
            return false;
        return true;
    }
public:
    bool isSymmetric(TreeNode *root) 
    {
        if (root == NULL) return true;
        return isSym(root->left, root->right);
    }
};

int main()
{
    return 0;
}

File: JavaSolution/src/SymmetricTree.java
/**
 * Created by xyunpeng on 7/6/15.
 */
public class SymmetricTree {
    public boolean isSymmetric(TreeNode root) {
        if (root == null)
            return true;
        else
            return helper(root.left, root.right);
    }

    public boolean helper(TreeNode left, TreeNode right) {
        if (left == null && right == null)
            return true;
        if (left == null || right == null)
            return false;
        if (left.val != right.val)
            return false;
        else
            return helper(left.left, right.right) && helper(right.left, left.right);
    }
}
